src/server/replica.cc

```diff
@@ -259,41 +214,61 @@ void Replica::ReplicateFb() {
       VLOG(1) << "Replica greet ok";
     }
 
+    // 4. Start stable state sync.
     DCHECK(state_mask_ & R_SYNC_OK);
 
-    // Start consuming the replication stream.
-
-    if (master_context_.dfly_session_id.empty())
-      ec = ConsumeRedisStream();
-    else
+    if (HasDflyMaster())
       ec = ConsumeDflyStream();
+    else
+      ec = ConsumeRedisStream();
 
     LOG_IF(ERROR, !FiberSocketBase::IsConnClosed(ec)) << "Replica socket error " << ec;
-    state_mask_ &= ~R_SYNC_OK;  //
+    state_mask_ &= ~R_SYNC_OK;
   }
 
-  VLOG(1) << "Replication fiber finished";
+  VLOG(1) << "Main replication fiber finished";
 }
 
-error_code Replica::Greet() {
-  base::IoBuf io_buf{128};
+error_code Replica::ConnectSocket() {
+  sock_.reset(ProactorBase::me()->CreateSocket());
 
-  ReqSerializer serializer{sock_.get()};
+  char ip_addr[INET6_ADDRSTRLEN];
+  int resolve_res = ResolveDns(master_context_.host, ip_addr);
+  if (resolve_res != 0) {
+    LOG(ERROR) << "Dns error " << gai_strerror(resolve_res) << ", host: " << master_context_.host;
+    return make_error_code(errc::host_unreachable);
+  }
+
+  master_context_.endpoint = {ip::make_address(ip_addr), master_context_.port};
+
+  /* These may help but require additional field testing to learn.
+   int yes = 1;
+   CHECK_EQ(0, setsockopt(sock_->native_handle(), IPPROTO_TCP, TCP_NODELAY, &yes, sizeof(yes)));
+   CHECK_EQ(0, setsockopt(sock_->native_handle(), SOL_SOCKET, SO_KEEPALIVE, &yes, sizeof(yes)));
+
+   int intv = 15;
+   CHECK_EQ(0, setsockopt(sock_->native_handle(), IPPROTO_TCP, TCP_KEEPIDLE, &intv, sizeof(intv)));
+
+   intv /= 3;
+   CHECK_EQ(0, setsockopt(sock_->native_handle(), IPPROTO_TCP, TCP_KEEPINTVL, &intv, sizeof(intv)));
 
+   intv = 3;
+   CHECK_EQ(0, setsockopt(sock_->native_handle(), IPPROTO_TCP, TCP_KEEPCNT, &intv, sizeof(intv)));
+  */
+
+  return sock_->Connect(master_context_.endpoint);
+}
+
+error_code Replica::Greet() {
   parser_.reset(new RedisParser{false});
```

Comment:
We create the parser in Greet for the main mode and StartAsDflyFlow separately... Can't we do this on startup? Or is this a "wasted init" in case of an error?

Reply:
I do not care about malloc. I think I need to reset parser state anyway in case it was used before (Greet has been called from the loop)